src/main/scala/mesosphere/marathon/api/v2/GroupNormalization.scala

```diff
@@ -1,21 +1,137 @@
 package mesosphere.marathon
 package api.v2
 
-import mesosphere.marathon.state.PathId
+import mesosphere.marathon.raml.{AppVisitor, GroupUpdate, GroupUpdateVisitor}
+import mesosphere.marathon.state.{AbsolutePathId, PathId, RootGroup}
+import mesosphere.mesos.ResourceMatcher.Role
+
+import scala.annotation.tailrec
+
+/**
+  * Visitor that normalizes a root group update.
+  *
+  * @param conf The [[MarathonConf]].
+  */
+case class RootGroupVisitor(conf: MarathonConf) extends GroupUpdateVisitor {
+
+  override def visit(thisGroup: GroupUpdate): GroupUpdate = thisGroup.copy(enforceRole = Some(false))
+
+  override def childGroupVisitor(): GroupUpdateVisitor = TopLevelGroupVisitor(conf)
+
+  override def appVisitor(): AppVisitor = AppNormalizeVisitor(conf, conf.mesosRole())
+}
+
+/**
+  * Visitor that normalizes a top-level group update, ie an update for a group directly under root eg `/prod`.
+  *
+  * @param conf The [[MarathonConf]].
+  */
+case class TopLevelGroupVisitor(conf: MarathonConf) extends GroupUpdateVisitor {
+  var defaultRole: Role = conf.mesosRole()
+
+  override def visit(thisGroup: raml.GroupUpdate): raml.GroupUpdate = {
+    val enforceRole = thisGroup.enforceRole.getOrElse {
+      conf.groupRoleBehavior() match {
+        case GroupRoleBehavior.Off => false
+        case GroupRoleBehavior.Top => true
+      }
+    }
+    if (enforceRole) defaultRole = PathId(thisGroup.id.get).root
+
+    thisGroup.copy(enforceRole = Some(enforceRole))
+  }
+
+  override def childGroupVisitor(): GroupUpdateVisitor = ChildGroupVisitor(conf, defaultRole)
+
+  override def appVisitor(): AppVisitor = AppNormalizeVisitor(conf, defaultRole)
+}
+
+/**
+  * Visitor that normalizes an update for a group that is not root or a top-level group. See
+  * [[RootGroupVisitor]] and [[TopLevelGroupVisitor]] for these cases.
+  *
+  * @param conf The [[MarathonConf]].
+  * @param defaultRole The default Mesos role for all apps in this group.
+  */
+case class ChildGroupVisitor(conf: MarathonConf, defaultRole: Role) extends GroupUpdateVisitor {
+
+  override def visit(thisGroup: GroupUpdate): GroupUpdate = {
+    if (thisGroup.enforceRole.isEmpty) thisGroup.copy(enforceRole = Some(false))
+    else thisGroup
+  }
+
+  override def childGroupVisitor(): GroupUpdateVisitor = this
+
+  override def appVisitor(): AppVisitor = AppNormalizeVisitor(conf, defaultRole)
```

Comment:
since these are side-effect-free I'd be inclined to omit ()

Reply:
Is this a common pattern? The interface might change but I'll make the change.